### ISSUE TITLE
Missing check for existence of method log_deprecation before calling it

### ISSUE THREAD

**Issue by jmanero-r7** [chef-cookbooks/compat_resource#7]
The change committed in 3139258 does not check for the existence of method `log_deprecation` before attempting to fetch it from `superclass`, resulting in exceptions when `superclass.log_deprecation` is not defined.

In this case, `superclass` is actually `Class`, which most certainly does not have a `log_deprecation` method. I'm not sure that you're ever actually going to have a `super` of `log_deprecation` for this module because of the way it's being instantiated and injected into `class Chef`'s class-methods. As far as I can tell, `module ChefCompatDeprecation` never gets injected into a class that has `class Chef` as a super. Rather, you may need to use an alias idiom like 

```
if respond_to?(:log_deprecation) ## && Maybe some other conditions
  alias_method :__to_another_place, :log_deprecation
else
  def __to_another_place; end
end

def log_deprecation(...)
    ## Do things and/or call `__to_another_place`
end
```

### Environment:

* Chef `chef_12.5.1-1_amd64`
* Ubuntu 14.04
* Cookbook `compat_resource@12.5.11`
* Cookbook `python@1.4.6`

### Stack Trace

```
================================================================================
Recipe Compile Error in [...]
================================================================================
NameError
---------
undefined method `log_deprecation' for class `Class'
Cookbook Trace:
---------------
/var/chef/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef.rb:6:in `method'
/var/chef/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef.rb:6:in `log_deprecation'
/var/chef/cookbooks/python/recipes/pip.rb:52:in `block in from_file'
/var/chef/cookbooks/python/recipes/pip.rb:50:in `from_file'
[...]

Relevant File Content:
----------------------
/var/chef/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef.rb:

1:  class Chef
2:    NOT_PASSED = Object.new if !defined?(NOT_PASSED)
3:    # Earlier versions of Chef didn't have this message
4:    module ChefCompatDeprecation
5:      def log_deprecation(message, location=nil)
6>>       if superclass.method(:log_deprecation).owner != ChefCompatDeprecation
7:          if !location
8:            # Pick the first caller that is *not* part of the Chef or ChefCompat gem,
9:            # that's the thing the user wrote.
10:            chef_compat_gem_path = File.expand_path("../../..", __FILE__)
11:            chef_gem_path = File.expand_path("../..",::Chef::Resource.instance_method(:initialize).source_location[0])
12:            caller(0..10).each do |c|
13:              if !c.start_with?(chef_gem_path) && !c.start_with?(chef_compat_gem_path)
14:                location = c
15:                break

[2015-10-28T23:20:01+00:00] ERROR: Running exception handlers
[2015-10-28T23:20:01+00:00] ERROR: Exception handlers complete
```

**Comment by sunggun-yu** [comment]
+1 same here.

with tomcat and rabbitmq cookbook.
trouble with compat_resource 12.5.11.

I could avoid this issue by adding `depends 'compat_resource', '= 12.5.10'` on metadata.rb file.
```
NameError
---------
undefined method `log_deprecation' for class `Class'

Cookbook Trace:
---------------
  /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef.rb:6:in `method'
  /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef.rb:6:in `log_deprecation'
  /var/chef/cache/cookbooks/tomcat/recipes/default.rb:65:in `block in from_file'
  /var/chef/cache/cookbooks/tomcat/recipes/default.rb:63:in `from_file'
  /var/chef/cache/cookbooks/hermes/recipes/tomcat.rb:19:in `from_file'

Relevant File Content:
----------------------
/var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef.rb:

  1:  class Chef
  2:    NOT_PASSED = Object.new if !defined?(NOT_PASSED)
  3:    # Earlier versions of Chef didn't have this message
  4:    module ChefCompatDeprecation
  5:      def log_deprecation(message, location=nil)
  6>>       if superclass.method(:log_deprecation).owner != ChefCompatDeprecation
  7:          if !location
  8:            # Pick the first caller that is *not* part of the Chef or ChefCompat gem,
  9:            # that's the thing the user wrote.
 10:            chef_compat_gem_path = File.expand_path("../../..", __FILE__)
 11:            chef_gem_path = File.expand_path("../..",::Chef::Resource.instance_method(:initialize).source_location[0])
 12:            caller(0..10).each do |c|
 13:              if !c.start_with?(chef_gem_path) && !c.start_with?(chef_compat_gem_path)
 14:                location = c
 15:                break
```

**Comment by raul782** [comment]
Just hit the same issue with an old cookbook and compat resource set to 12.5.11

**Comment by shanesveller** [comment]
I'm seeing this by way of the consul and/or consul-template cookbooks.

**Comment by rgriffith** [comment]
Just ran into the exact same error during a chef-client run on a CentOS 6.5 node, seems to be coming from including latest iptables cookbook (2.0.0).

**Comment by kpmueller** [comment]
Also seeing this issue.

**Comment by ksexton** [comment]
Seeing this as well, compat_resource pinning in metadata.rb from @sunggun-yu seems to resolve.

**Comment by aaron-lane** [comment]
This is essentially the same problem as #3

**Comment by whiteadam** [comment]
+1

**Comment by jlspies1** [comment]
+1 - and I wonder if this won't hit a vast majority of users, I'm seeing it in particular with a New Relic cookbook. If we have to pin to version 12.5.10 on all of our cookbooks (or test all of them individually), this will be time consuming.

**Comment by donwitti** [comment]
+1

**Comment by christinedraper** [comment]
+1

**Comment by Chili-Man** [comment]
+1

**Comment by levmichael3** [comment]
+1.
Only failing on Chef: 12.5.1
I have other nodes with Chef: 12.4.0 which stay happy.

**Comment by andrefreitas** [comment]
+1, please fix this

**Comment by Rockstar04** [comment]
Just hit this issue as well.

What is the plan to fix this?? What can I do to help?

**Comment by Sliim** [comment]
+1

**Comment by jkeiser** [comment]
Sorry, missed this over the weekend. Fixing now.

**Comment by jkeiser** [comment]
Found out why this regressed, at least; 12.5.1 was not running in travis due to bundler cache. I've added it.

**Comment by kpmueller** [comment]
thank you!

**Comment by jkeiser** [comment]
This is now released as 12.5.13, and the tests should catch instances like this in the future. Sorry about the inconvenience!
